File: Editorial/70_Climbing_Stairs/CSharp/5_Binets_Method.cs
/* C# */

public class Solution {
    public int ClimbStairs(int n) {
        int[][] q = new int [2][] { new int[2] { 1, 1 }, new int[2] { 1, 0 } };
        int[][] res = Pow(q, n);
        return res[0][0];
    }

    public int[][] Pow(int[][] a, int n) {
        int[][] ret =
            new int [2][] { new int[2] { 1, 0 }, new int[2] { 0, 1 } };
        while (n > 0) {
            if ((n & 1) == 1) {
                ret = Multiply(ret, a);
            }

            n >>= 1;
            a = Multiply(a, a);
        }

        return ret;
    }

    public int[][] Multiply(int[][] a, int[][] b) {
        int[][] c = new int [2][] { new int[2] { 0, 0 }, new int[2] { 0, 0 } };
        for (int i = 0; i < 2; i++) {
            for (int j = 0; j < 2; j++) {
                c[i][j] = a[i][0] * b[0][j] + a[i][1] * b[1][j];
            }
        }

        return c;
    }
}

File: Editorial/70_Climbing_Stairs/Python3/5_Binets_Method.py
# Python3
class Solution:
    def climbStairs(self, n: int) -> int:
        q = [[1, 1], [1, 0]]
        res = self.pow(q, n)
        return res[0][0]

    def pow(self, a: [[int]], n: int) -> [[int]]:
        ret = [[1, 0], [0, 1]]
        while n > 0:
            if (n & 1) == 1:
                ret = self.multiply(ret, a)
            n >>= 1
            a = self.multiply(a, a)
        return ret

    def multiply(self, a: [[int]], b: [[int]]) -> [[int]]:
        c = [[0, 0], [0, 0]]
        for i in range(2):
            for j in range(2):
                c[i][j] = a[i][0] * b[0][j] + a[i][1] * b[1][j]
        return c
